src/beancount.pest

```diff
@@ -1,165 +1,168 @@
 
 //// Date primitives
-year = { ASCII_DIGIT{4} }
-month = { "0" ~ ASCII_DIGIT | "1" ~ '0'..'2' }
-day = { '0'..'2' ~ ASCII_DIGIT | "3" ~ '0'..'1' }
-date_separator = { "-" | "/" }
-date = { year ~ date_separator ~ month ~ date_separator ~ day }
+year = @{ ASCII_DIGIT{4} }
+month = @{ "0" ~ ASCII_DIGIT | "1" ~ '0'..'2' }
+day = @{ '0'..'2' ~ ASCII_DIGIT | "3" ~ '0'..'1' }
+date_separator = @{ "-" | "/" }
+date = ${ year ~ date_separator ~ month ~ date_separator ~ day }
 
 //// Number primitives
 num = @{ int ~ ("." ~ ASCII_DIGIT*)? }
-    int = { ("+" | "-")? ~ ( ASCII_DIGIT{4,} | separated_int ) }
-    separated_int = { ASCII_DIGIT{1,3} ~ ( "," ~ ASCII_DIGIT{3} )* }
+    int = @{ ("+" | "-")? ~ ( ASCII_DIGIT{4,} | separated_int ) }
+    separated_int = @{ ASCII_DIGIT{1,3} ~ ( "," ~ ASCII_DIGIT{3} )* }
 operation = _{ add | subtract | multiply | divide }
-    add      = { "+" }
-    subtract = { "-" }
-    multiply = { "*" }
-    divide   = { "/" }
+    add      = @{ "+" }
+    subtract = @{ "-" }
+    multiply = @{ "*" }
+    divide   = @{ "/" }
 num_expr = { term ~ (operation ~ term)* }
 term = _{ num | "(" ~ num_expr ~ ")" }
-amount = { num_expr ~ ws+ ~ commodity }
+amount = { num_expr ~ commodity }
 
 //// String primitives
-double_quote = { "\"" }
-quoted_str = { double_quote ~ inner_quoted_str ~ double_quote }
+double_quote = @{ "\"" }
+quoted_str = @{ double_quote ~ inner_quoted_str ~ double_quote }
 inner_quoted_str = @{ quoted_char* }
-quoted_char = { escape_sequence | ( !"\"" ~ ANY ) }
-escape_sequence = { "\\" ~ ANY }
-valid_non_letter_commodity_char = { "'" |  "_" | "-" | "." }
-commodity_trailing = { valid_non_letter_commodity_char ~ &commodity_trailing | (ASCII_ALPHA_UPPER | ASCII_DIGIT) }
+quoted_char = @{ escape_sequence | ( !"\"" ~ ANY ) }
+escape_sequence = @{ "\\" ~ ANY }
+valid_non_letter_commodity_char = @{ "'" |  "_" | "-" | "." }
+commodity_trailing = @{ valid_non_letter_commodity_char ~ &commodity_trailing | (ASCII_ALPHA_UPPER | ASCII_DIGIT) }
 commodity = @{ ASCII_ALPHA_UPPER ~ commodity_trailing{1, 23} }
+commodity_list = @{ commodity ~ ("," ~ commodity)* }
 
 //// Account primitives
-account_type = { UPPERCASE_LETTER ~ (LETTER | DECIMAL_NUMBER | "-")* }
-account_name_piece = { ":" ~ (UPPERCASE_LETTER | DECIMAL_NUMBER) ~ (LETTER | DECIMAL_NUMBER | "-")* }
+account_type = @{ UPPERCASE_LETTER ~ (LETTER | DECIMAL_NUMBER | "-")* }
+account_name_piece = @{ ":" ~ (UPPERCASE_LETTER | DECIMAL_NUMBER) ~ (LETTER | DECIMAL_NUMBER | "-")* }
 account = @{ account_type ~ account_name_piece+ }
-tag_name = { ( ASCII_ALPHANUMERIC | "-" | "_" | "/" | "." )+ }
-link = { "^" ~ tag_name }
-tag = { "#" ~ tag_name }
-tags_links = { (link | tag) ~ (ws+ ~ (link | tag))* }
+tag_name = @{ ( ASCII_ALPHANUMERIC | "-" | "_" | "/" | "." )+ }
+link = ${ "^" ~ tag_name }
+tag = ${ "#" ~ tag_name }
+tags_links = { (link | tag)+ }
 
 //// Directives
 
+// * Org mode title
+org_mode_title = @{ WHITESPACE* ~ "*" ~ (!NEWLINE ~ ANY)* ~ eol }
```

Comment:
Ahh, cool!  It might be nice if we could simply skip this expression / line vs. producing a token to match the behavior of Beancount proper: https://bitbucket.org/blais/beancount/src/8e65487af0c0d2017b7f5e38686da5fd52aa3652/beancount/parser/lexer.l?at=default&fileviewer=file-view-default#lexer.l-181

Although probably not the end of the world if we do produce a token for it.  Looks like silent _and_ atomic expressions are currently not available in Pest, at least not in the naive manner of `_@{ ... }`.  https://github.com/pest-parser/pest/issues/118 leads to a solution that may eventually come to Pest.

Also, I'm unable to get `bean-check` to accept something like this:

```
$ cat org-mode-test.beancount 
 * my_org_mode_title

1990-12-02 commodity HOOL
  name: "Hooli Common shares."
```

```
$ bean-check org-mode-test.beancount 
/Users/me/projects/beancount-sandbox/org-mode-test.beancount:1:       syntax error, unexpected ASTERISK

/Users/me/projects/beancount-sandbox/org-mode-test.beancount:1:       Invalid token: 'my_org_mode_title'
```

...but I'm also not actually using org-mode and instead just inserted the asterisk and text myself.  Perhaps I'm doing something wrong here.  I don't use Emacs myself 🙂

Reply:
Oh, looks like Beancount doesn't allow any whitespace before the asterisk. I thought it did. I've fixed that.